design-proposals/klip-36-grace-stream-stream-joins.md

```diff
@@ -0,0 +1,102 @@
+# KLIP 36 - GRACE period for stream-stream joins
+
+**Author**: agavra | 
+**Release Target**: 0.14 | 
+**Status**: In Discussion | 
+**Discussion**: TBD
+
+**tl;dr:** _Support controlling the GRACE period for a stream-stream join to improve disk space
+utilization for small windows_
+           
+## Motivation and background
+
+As described in https://github.com/confluentinc/ksql/issues/6152, the ability to specify
+the grace period for stream-stream joins allows more control over the disk space utilization.
+This is because a window needs to stay on disk until the grace period has elapsed in order to
+ensure that any events that come in timestamped during an old window can be processed as long
+as the grace period is active. This means that for windows dramatically smaller than the grace
+period (e.g. a minute) we need to keep that level of granularity for the default 24 hour window.
+
+## What is in scope
+
+- Syntax for specifying grace period within stream-stream joins
+
+## What is not in scope
+
+- N/A
+
+## Value/Return
+
+Enables onboarding use cases that perform stream-stream joins on high cardinality, high throughput
+datasets with small windows without exhausting too much disk space.
+
+## Public APIS
+
+Stream-stream joins require use of `WITHIN` clauses:
+
+```
+JOIN <stream> WITHIN <time unit> ON <condition>
+```
+
+This will be expanded to support a more complex `withinExpression` that leverages the same
+syntax as aggregation windows:
+
+```
+JOIN <stream> WITHIN (SIZE <time unit>, GRACE PERIOD <time unit>) ON <condition>
+```
+
+The old syntax will still be supported for backwards compatibility.
```

Comment:
Are there plans to remove the old syntax eventually?

For my own education: my understanding was that `GRACE PERIOD` is optional in aggregations? Is this correct? Would it be optional for stream-stream joins, too?

For Kafka Streams, we had a discussions about making it a mandatory parameter.

Reply:
> Are there plans to remove the old syntax eventually?

I think it would make sense to remove the old syntax for 1.0.

> Would it be optional for stream-stream joins, too?

Yes (this is necessary to allow the old syntax as well)

Reply:
Well, even if we make it required, we could still set 24h if the old syntax is used, when translating the old to the new internally?

Reply:
Curious to hear what other people think? Should GRACE PERIOD be optional? The current default of 24h is rather "random" and using a default of zero implies bad out-of-the-box behavior with regard to out-of-order data -- we would discard out-of-order records very aggressively as later records.

Reply:
With reference to recent discussions around supporting older versions of the syntax for longer...  we'd likely want to support the old syntax for a while.